src/core/api/robot.go

```diff
@@ -0,0 +1,197 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"fmt"
+	"github.com/goharbor/harbor/src/common/dao"
+	"github.com/goharbor/harbor/src/common/models"
+	"net/http"
+	"strconv"
+)
+
+// User this prefix to distinguish harbor user,
+// The prefix contains a specific character($), so it cannot be registered as a harbor user.
+const robotPrefix = "robot$"
+
+// RobotAPI ...
+type RobotAPI struct {
+	BaseController
+	project *models.Project
+	robot   *models.Robot
+}
+
+// Prepare ...
+func (r *RobotAPI) Prepare() {
+	r.BaseController.Prepare()
+	method := r.Ctx.Request.Method
+
+	if !r.SecurityCtx.IsAuthenticated() {
+		r.HandleUnauthorized()
+		return
+	}
+
+	pid, err := r.GetInt64FromPath(":pid")
+	if err != nil || pid <= 0 {
+		text := "invalid project ID: "
+		if err != nil {
+			text += err.Error()
+		} else {
+			text += fmt.Sprintf("%d", pid)
+		}
+		r.HandleBadRequest(text)
+		return
+	}
+	project, err := r.ProjectMgr.Get(pid)
+	if err != nil {
+		r.ParseAndHandleError(fmt.Sprintf("failed to get project %d", pid), err)
+		return
+	}
+	if project == nil {
+		r.HandleNotFound(fmt.Sprintf("project %d not found", pid))
+		return
+	}
+	r.project = project
+
+	if method == http.MethodPut || method == http.MethodDelete {
+		id, err := r.GetInt64FromPath(":id")
+		if err != nil || id <= 0 {
+			r.HandleBadRequest("invalid robot ID")
+			return
+		}
+
+		robot, err := dao.GetRobotByID(id)
+		if err != nil {
+			r.HandleInternalServerError(fmt.Sprintf("failed to get robot %d: %v", id, err))
+			return
+		}
+
+		if robot == nil {
+			r.HandleNotFound(fmt.Sprintf("robot %d not found", id))
+			return
+		}
+
+		r.robot = robot
+	}
+
+	if !(r.Ctx.Input.IsGet() && r.SecurityCtx.HasReadPerm(pid) ||
+		r.SecurityCtx.HasAllPerm(pid)) {
+		r.HandleForbidden(r.SecurityCtx.GetUsername())
+		return
+	}
+
+}
+
+// Post ...
+func (r *RobotAPI) Post() {
+	var robotReq models.RobotReq
+	r.DecodeJSONReq(&robotReq)
+
+	createdName := robotPrefix + robotReq.Name
+
+	robot := models.Robot{
+		Name:        createdName,
+		Description: robotReq.Description,
+		ProjectID:   r.project.ProjectID,
+		// TODO: use token service to generate token per access information
+		Token: "this is a placeholder",
+	}
+
+	id, err := dao.AddRobot(&robot)
+	if err != nil {
+		if err == dao.ErrDupRows {
+			r.HandleConflict()
+			return
+		}
+		r.HandleInternalServerError(fmt.Sprintf("failed to create robot account: %v", err))
+		return
+	}
+
+	robotRep := models.RobotRep{
+		Name:  robot.Name,
+		Token: robot.Token,
+	}
+
+	r.Redirect(http.StatusCreated, strconv.FormatInt(id, 10))
+	r.Data["json"] = robotRep
+	r.ServeJSON()
+}
+
+// List list all the robots of a project
+func (r *RobotAPI) List() {
+	query := models.RobotQuery{
+		ProjectID: r.project.ProjectID,
+	}
+
+	count, err := dao.CountRobot(&query)
+	if err != nil {
+		r.HandleInternalServerError(fmt.Sprintf("failed to count robots %v", err))
```

Comment:
Can we change message `failed to count robots` to another one? It's too "backend" and not obvious relationship with `List` function.